horovod/torch/sync_batch_norm.py

```diff
@@ -0,0 +1,194 @@
+        input = input.contiguous()
+
+        size = input.numel() // input.size(1)
+        count = torch.tensor([size])
+
+        # calculate mean/invstd for input.
+        mean, invstd = torch.batch_norm_stats(input, eps)
+
+        count_handle = allgather_async(count.unsqueeze(0), name='sync_batch_norm.count')
+        mean_handle = allgather_async(mean.unsqueeze(0), name='sync_batch_norm.mean')
+        invstd_handle = allgather_async(invstd.unsqueeze(0), name='sync_batch_norm.invstd')
+
+        # wait on the async communication to finish
+        count_all = synchronize(count_handle)
+        mean_all = synchronize(mean_handle)
+        invstd_all = synchronize(invstd_handle)
+
+        if _SYNC_BN_V2:
+            counts_for_bngswc = count_all.view(-1).float().to(input.device)
+        else:
+            # backwards compatibility
+            counts_for_bngswc = count_all.view(-1).tolist()
+
+        # calculate global mean & invstd
+        mean, invstd = torch.batch_norm_gather_stats_with_counts(
+            input,
+            mean_all,
+            invstd_all,
+            running_mean,
+            running_var,
+            momentum,
+            eps,
+            counts_for_bngswc
+        )
+
+        self.save_for_backward(input, weight, mean, invstd, count_all)
+
+        # apply element-wise normalization
+        return torch.batch_norm_elemt(input, weight, bias, mean, invstd, eps)
+
+    @staticmethod
+    def backward(self, grad_output):
+        grad_output = grad_output.contiguous()
+        saved_input, weight, mean, invstd, count_all = self.saved_tensors
+        need_input_grad, need_weight_grad, need_bias_grad = self.needs_input_grad[0:3]
+
+        # calculate local stats as well as grad_weight / grad_bias
+        sum_dy, sum_dy_xmu, grad_weight, grad_bias = torch.batch_norm_backward_reduce(
+            grad_output,
+            saved_input,
+            mean,
+            invstd,
+            weight,
+            need_input_grad,
+            need_weight_grad,
+            need_bias_grad
+        )
+
+        if need_input_grad:
+            # synchronizing stats used to calculate input gradient.
+            sum_dy_handle = allreduce_async(sum_dy, op=Sum, name='sync_batch_norm.sum_dy')
+            sum_dy_xmu_handle = allreduce_async(sum_dy_xmu, op=Sum, name='sync_batch_norm.sum_dy_xmu')
+
+            # wait on the async communication to finish
+            sum_dy = synchronize(sum_dy_handle)
+            sum_dy_xmu = synchronize(sum_dy_xmu_handle)
+
+            if _SYNC_BN_V2:
+                mean_dy = sum_dy / count_all.sum()
+                mean_dy_xmu = sum_dy_xmu / count_all.sum()
+            else:
+                # before 1.6.0, sum_dy was sum of means from every worker, so we just 
+                # need to divide it by number of workers
+                mean_dy = sum_dy / size()
+                mean_dy_xmu = sum_dy_xmu / size()
```

Comment:
I think we should divide by `count_all.sum()` for both pytorch 1.5 and 1.6.
 
For pytorch 1.5, right now if I comment out `_run_bn` code path (not using torch.nn.BatchNorm) and run the following on `v0.19.3` locally:
```
     torch.cuda.manual_seed(2020)
     bn = torch.nn.BatchNorm2d(10).cuda()
     bn_hvd = SyncBatchNorm(10).cuda()

     x = torch.rand(3, 10, 8, 8).cuda()
     x1 = x.clone().requires_grad_()
     x2 = x.clone().requires_grad_()
     y = bn(x1)
     y_hvd = bn_hvd(x2)

     y.sum().backward()
     y_hvd.sum().backward()
     print((x1.grad - x2.grad).abs().sum())
```
I got 
```
tensor(1298650., device='cuda:0')
```
But if  if we divide by `count_all.sum()` here, I got:
```
tensor(0.0004, device='cuda:0')
```

Reply:
Great catch, I've raised #1980 to address this.  Apparently they made breaking changes in both 1.5.0 and 1.6.0.